clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -706,9 +708,13 @@ public Node leastLoadedNode(long now) {
     }
 
     public static AbstractResponse parseResponse(ByteBuffer responseBuffer, RequestHeader requestHeader) {
-        Struct responseStruct = parseStructMaybeUpdateThrottleTimeMetrics(responseBuffer, requestHeader, null, 0);
-        return AbstractResponse.parseResponse(requestHeader.apiKey(), responseStruct,
-                requestHeader.apiVersion());
+        try {
+            Struct responseStruct = parseStructMaybeUpdateThrottleTimeMetrics(responseBuffer, requestHeader, null, 0);
+            return AbstractResponse.parseResponse(requestHeader.apiKey(), responseStruct,
+                    requestHeader.apiVersion());
+        } catch (BufferUnderflowException e) {
+            throw new SchemaException("Buffer underflow while parsing response for request with header " + requestHeader, e);
```

Comment:
Was this change triggered by a test failure?

Reply:
Yes. The new schema causes a `BufferUnderflowException` when in the  `REAUTH_RECEIVE_HANDSHAKE_OR_OTHER_RESPONSE` state in `SaslClientAuthenticator`. This class expects to be able to catch parsing errors in order to change the handling.